File: C++_version/PriorityNonPreemptive.cpp
#include "PriorityNonPreemptive.hpp"
#include <iostream>
#include <algorithm>

PriorityNonPreemptive::PriorityNonPreemptive(std::vector<Process> &process) : Scheduler(process) {}
PriorityNonPreemptive::PriorityNonPreemptive() {}

int PriorityNonPreemptive::schedule()
{
    int id;

    // Only search for a new process if the current one finished
    if (running_id == 0)
    {
        Process* best_process = nullptr;

        for (auto& proc : process_queue)
        {
            bool has_arrived = proc.getArrival_time() <= current_time;
            bool not_finished = proc.getExecution_time() < proc.getBurst_time();

            if (has_arrived && not_finished)
            {
                if (best_process == nullptr || proc.getPriority() < best_process->getPriority())
                {
                    best_process = &proc;
                }
            }
        }

        if (best_process != nullptr)
        {
            running_process = std::find_if(process_queue.begin(), process_queue.end(),
                                           [best_process](const Process& p) { return p.getID() == best_process->getID(); });

            running_id = best_process->getID();
        }
    }

    id = running_id;

    if (running_id != 0 && running_process != process_queue.end())
    {
        running_process->setExecution_time(running_process->getExecution_time() + 1);

        if (running_process->getExecution_time() == running_process->getBurst_time())
        {
            int finish_time = current_time + 1;
            int arrival = running_process->getArrival_time();
            int burst = running_process->getBurst_time();

            running_process->setwaiting_time(finish_time - arrival - burst);
            running_process->setturnaround_time(finish_time - arrival);
            running_id = 0;
        }
    }

    current_time++;
    return id;
}

File: java_version/PriorityNonPreemptive.java
import java.util.List;
import java.util.Queue;
import java.util.ArrayList;

public class PriorityNonPreemptive extends Scheduler {
    private List<Process> completedProcesses = new ArrayList<>();

    public PriorityNonPreemptive(List<Process> processes) {
        super(processes);
    }

    public PriorityNonPreemptive() {}

    @Override
    public int schedule() {
        int processId = runningProcess != null ? runningProcess.getId() : 0;

        if (runningProcess == null) {
            Process selected = null;
            for (Process p : processQueue) {
                if (p.getArrivalTime() <= currentTime && p.getExecutionTime() < p.getBurstTime()) {
                    if (selected == null || p.getPriority() < selected.getPriority()) {
                        selected = p;
                    }
                }
            }
            runningProcess = selected;
            if (runningProcess != null) {
                processId = runningProcess.getId();
            }
        }

        if (runningProcess != null) {
            runningProcess.setExecutionTime(runningProcess.getExecutionTime() + 1);
            if (runningProcess.getExecutionTime() == runningProcess.getBurstTime()) {
                runningProcess.setWaitingTime(currentTime + 1 - runningProcess.getArrivalTime() - runningProcess.getBurstTime());
                runningProcess.setTurnaroundTime(currentTime + 1 - runningProcess.getArrivalTime());
                completedProcesses.add(runningProcess);
                runningProcess = null;
            }
        }

        currentTime++;
        return processId;
    }

    public void addProcess(int processId, int priority) {
        // Assuming the Process class does not take priority in the constructor
        Process newProcess = new Process(processId, 0, 0); // Default burst and arrival times
        newProcess.setPriority(priority); // Set priority separately if supported
        processQueue.add(newProcess);
    }
    
    public synchronized void addProcessWithPriority(int burstTime, int arrivalTime, int priority) {
        // Assuming the Process class does not take priority in the constructor
        Process newProcess = new Process(processQueue.size() + 1, burstTime, arrivalTime); // Create process
        newProcess.setPriority(priority); // Set priority separately if supported
        processQueue.add(newProcess);
    }


    public String getGanttChart() {
        StringBuilder ganttChart = new StringBuilder();
        for (Process p : completedProcesses) {
            ganttChart.append("P").append(p.getId()).append(" ");
        }
        return ganttChart.toString().trim();
    }

  
    public void schedule(int timeQuantum, Queue<Integer> processQueue) {
        throw new UnsupportedOperationException("Time quantum scheduling is not supported in non-preemptive priority scheduling.");
    }

    public double getAverageTurnaroundTime() {
        double totalTurnaroundTime = 0;
        for (Process p : completedProcesses) {
            totalTurnaroundTime += p.getTurnaroundTime();
        }
        return completedProcesses.isEmpty() ? 0.0 : totalTurnaroundTime / completedProcesses.size();
    }


    public double getAverageWaitingTime() {
        double totalWaitingTime = 0;
        for (Process p : completedProcesses) {
            totalWaitingTime += p.getWaitingTime();
        }
        return completedProcesses.isEmpty() ? 0.0 : totalWaitingTime / completedProcesses.size();
    }

  
    public int getRemainingBurst(int processId) {
        for (Process p : processQueue) {
            if (p.getId() == processId) {
                return p.getBurstTime() - p.getExecutionTime();
            }
        }
        return 0;
    }

    public String getCurrentProcessName() {
        return runningProcess != null ? "P" + runningProcess.getId() : "Idle";
    }
}
